Jetty 12.1 brotli linking problem

**Jetty version(s)**

12.1.0

**Jetty Environment**

**Java version/vendor** `(use: java -version)`

24.0.2

**OS type/version**

Linux 

```
[root@presto-worker-1 /]# /lib64/libc.so.6
GNU C Library (GNU libc) stable release version 2.34.
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 11.5.0 20240719 (Red Hat 11.5.0-5).
libc ABIs: UNIQUE ABSOLUTE
For bug reporting instructions, please see:
<https://www.gnu.org/software/libc/bugs.html>.
```

```
Linux presto-worker-1 6.14.10-orbstack-00291-g1b252bd3edea #1 SMP Sat Jun  7 02:45:18 UTC 2025 aarch64 aarch64 aarch64 GNU/Linux
```

**Description**

Brotli linking exception prevents server startup (this could happen in cases like: dynamic library cannot be loaded from the location that is mounted with noexec or brotli is linked against newer libc). It should be possible to apply brotli and any other compression conditionally, if it's available. Right now since the static blocks are used to check for native library availability, it's not possible to correctly intercept linking exception without the usage of reflection.

**How to reproduce?**

@sbordet it was mentioned in https://github.com/jetty/jetty.project/issues/13493#issuecomment-3210728203

```
presto-worker-2     | Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.UnsatisfiedLinkError: Failed to load Brotli native library [in thread "main"]
presto-worker-2     |   at com.aayushatharva.brotli4j.Brotli4jLoader.ensureAvailability(Brotli4jLoader.java:108)
presto-worker-2     |   at org.eclipse.jetty.compression.brotli.BrotliCompression.<clinit>(BrotliCompression.java:54)
```

Client has the same issue while using ServiceLoader to discover compressions:

```
presto-master       | Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.UnsatisfiedLinkError: Failed to load Brotli native library [in thread "main"]
presto-master       |   at com.aayushatharva.brotli4j.Brotli4jLoader.ensureAvailability(Brotli4jLoader.java:108)
presto-master       |   at org.eclipse.jetty.compression.brotli.BrotliCompression.<clinit>(BrotliCompression.java:54)
presto-master       |   at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
presto-master       |   at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1166)
presto-master       |   at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:341)
presto-master       |   at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:104)
presto-master       |   at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:143)
presto-master       |   at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:546)
presto-master       |   at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:496)
presto-master       |   at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
presto-master       |   at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:707)
presto-master       |   at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:672)
presto-master       |   at java.base/java.util.ServiceLoader$2.next(ServiceLoader.java:1256)
presto-master       |   at org.eclipse.jetty.util.ServiceLoaderSpliterator.tryAdvance(ServiceLoaderSpliterator.java:39)
presto-master       |   at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
presto-master       |   at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
presto-master       |   at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
presto-master       |   at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
presto-master       |   at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
presto-master       |   at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
presto-master       |   at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:636)
presto-master       |   at org.eclipse.jetty.client.HttpClient.doStart(HttpClient.java:243)
```

--- a/jetty-core/jetty-compression/jetty-compression-brotli/src/main/java/org/eclipse/jetty/compression/brotli/BrotliCompression.java
+++ b/jetty-core/jetty-compression/jetty-compression-brotli/src/main/java/org/eclipse/jetty/compression/brotli/BrotliCompression.java
@@ -49,17 +49,15 @@ public class BrotliCompression extends Compression
     private static final HttpField CONTENT_ENCODING = new PreEncodedHttpField(HttpHeader.CONTENT_ENCODING, ENCODING_NAME);
     private static final int DEFAULT_MIN_BROTLI_SIZE = 48;
 
-    static
-    {
-        Brotli4jLoader.ensureAvailability();
-    }
-
     private BrotliEncoderConfig defaultEncoderConfig = new BrotliEncoderConfig();
     private BrotliDecoderConfig defaultDecoderConfig = new BrotliDecoderConfig();
 
     public BrotliCompression()
     {
         super(ENCODING_NAME);
+        // Brotli4jLoader uses a static class block to link the
+        // native library, so it is thread-safe by definition.
+        Brotli4jLoader.ensureAvailability();
         setMinCompressSize(DEFAULT_MIN_BROTLI_SIZE);
     }
 

--- a/jetty-core/jetty-compression/jetty-compression-zstandard/src/main/java/org/eclipse/jetty/compression/zstandard/ZstandardCompression.java
+++ b/jetty-core/jetty-compression/jetty-compression-zstandard/src/main/java/org/eclipse/jetty/compression/zstandard/ZstandardCompression.java
@@ -22,6 +22,7 @@ import java.util.Objects;
 
 import com.github.luben.zstd.ZstdInputStreamNoFinalizer;
 import com.github.luben.zstd.ZstdOutputStreamNoFinalizer;
+import com.github.luben.zstd.util.Native;
 import org.eclipse.jetty.compression.Compression;
 import org.eclipse.jetty.compression.DecoderConfig;
 import org.eclipse.jetty.compression.DecoderSource;
@@ -61,6 +62,9 @@ public class ZstandardCompression extends Compression
     public ZstandardCompression()
     {
         super(ENCODING_NAME);
+        // Zstandard's Native uses a static class block to link
+        // the native library, so it is thread-safe by definition.
+        Native.load();
         setMinCompressSize(DEFAULT_MIN_ZSTD_SIZE);
     }
 

